File: Speculum.Api/Motor/Live/DeviceProfile.cs
using MessagePack;

namespace Speculum.Api.Motor.Live;

/// <summary>
/// Optional mobile/desktop emulation profile for StartSession / Resize.
/// MessagePack keys are camelCase for the web client.
/// </summary>
[MessagePackObject]
public sealed class DeviceProfile
{
    [Key("mobile")]
    public bool Mobile { get; init; }

    [Key("touch")]
    public bool Touch { get; init; }

    [Key("deviceScaleFactor")]
    public double DeviceScaleFactor { get; init; } = 1;

    [Key("maxTouchPoints")]
    public int MaxTouchPoints { get; init; }

    /// <summary>Stable profile id — not a free-form UA string from the client.</summary>
    [Key("userAgentProfile")]
    public string? UserAgentProfile { get; init; }

    [Key("screenOrientation")]
    public string? ScreenOrientation { get; init; }
}

File: sidecar/src/protocol/device-profile.ts
export type DeviceProfile = {
    mobile: boolean;
    touch: boolean;
    deviceScaleFactor: number;
    maxTouchPoints: number;
    userAgentProfile?: string;
    screenOrientation?: string;
};

export const DEFAULT_DEVICE_PROFILE: DeviceProfile = {
    mobile: false,
    touch: false,
    deviceScaleFactor: 1,
    maxTouchPoints: 0,
    userAgentProfile: 'desktop',
};

const MAX_DPR = 2;
const MAX_TOUCH_POINTS = 10;

export function normalizeDeviceProfile(raw: Partial<DeviceProfile> | undefined | null): DeviceProfile {
    if (!raw) return { ...DEFAULT_DEVICE_PROFILE };

    let dpr = Number(raw.deviceScaleFactor);
    if (!Number.isFinite(dpr) || dpr <= 0) dpr = 1;
    dpr = Math.min(MAX_DPR, Math.max(1, dpr));

    let maxTouchPoints = Number(raw.maxTouchPoints) || 0;
    if (maxTouchPoints < 0) maxTouchPoints = 0;
    if (maxTouchPoints > MAX_TOUCH_POINTS) maxTouchPoints = MAX_TOUCH_POINTS;

    const mobile = !!raw.mobile;
    let touch = !!raw.touch || mobile;
    if (touch && maxTouchPoints === 0) maxTouchPoints = 5;

    let userAgentProfile = raw.userAgentProfile;
    if (userAgentProfile !== 'mobile' && userAgentProfile !== 'desktop') {
        userAgentProfile = mobile ? 'mobile' : 'desktop';
    }

    return {
        mobile,
        touch,
        deviceScaleFactor: dpr,
        maxTouchPoints,
        userAgentProfile,
        screenOrientation: raw.screenOrientation,
    };
}
